packages/store/addon/-private/system/store/record-data-store-wrapper.ts

```diff
@@ -107,36 +110,46 @@ export default class RecordDataStoreWrapper implements IRecordDataStoreWrapper {
 
   inverseForRelationship(type: string, key: string): string | null {
     const modelClass = this._store.modelFor(type);
-    const definition = upgradeForInternal(this.relationshipsDefinitionFor(type)[key]);
+    const definition = this.relationshipsDefinitionFor(type)[key];
+    if (!definition) {
+      return null;
+    }
     if (CUSTOM_MODEL_CLASS) {
       if (definition.inverse !== undefined) {
         return definition.inverse;
       } else {
         //TODO add a test for this branch
-        if (!definition._inverseKey) {
-          return null;
+        if (metaIsRelationshipDefinition(definition)) {
+          return definition._inverseKey(this._store, modelClass);
         }
-        return definition._inverseKey(this._store, modelClass);
+        return null;
       }
     } else {
-      return definition._inverseKey(this._store, modelClass);
+      return (definition as RelationshipDefinition)._inverseKey(this._store, modelClass);
     }
   }
 
   inverseIsAsyncForRelationship(type: string, key: string): boolean {
     const modelClass = this._store.modelFor(type);
-    const definition = upgradeForInternal(this.relationshipsDefinitionFor(type)[key]);
+    const definition = this.relationshipsDefinitionFor(type)[key];
+    if (!definition) {
+      return false;
+    }
     if (CUSTOM_MODEL_CLASS) {
       if (definition.inverse === null) {
         return false;
       }
-      if (definition.inverseIsAsync !== undefined) {
-        return !!definition.inverseIsAsync;
-      } else {
+      if (((definition as unknown) as { inverseIsAsync?: boolean }).inverseIsAsync !== undefined) {
```

Comment:
Whats up with multiple casts?

Reply:
even though RelationshipDefinition implements RelationshipMeta typescript does not allow us to recast to it without first casting to `unknown` unless we change the return type of `relationshipsDefinitionFor` to be a union type. Since this is the only location that we leak our current implementation detail for meta on @ember-data/model it seemed better to leave the return type there alone.